definition_validator/validator.cc

```diff
@@ -189,6 +189,78 @@ void validateOverriding(const core::GlobalState &gs, core::SymbolRef method) {
     }
 }
 
+core::Loc getAncestorLoc(const core::GlobalState &gs, const ast::ClassDef *const classDef,
+                         const core::SymbolRef ancestor) {
+    for (const auto &anc : classDef->ancestors) {
+        const auto ancConst = ast::cast_tree<ast::ConstantLit>(anc.get());
+        if (ancConst == nullptr) {
+            continue;
+        }
+        if (ancConst->symbol.data(gs)->dealias(gs) == ancestor) {
+            return anc->loc;
+        }
+    }
+    for (const auto &anc : classDef->singletonAncestors) {
+        const auto ancConst = ast::cast_tree<ast::ConstantLit>(anc.get());
+        if (ancConst == nullptr) {
+            continue;
+        }
+        if (ancConst->symbol.data(gs)->dealias(gs) == ancestor) {
+            return anc->loc;
+        }
+    }
+    // give up
+    return classDef->loc;
+}
+
+void validateFinalAncestorHelper(const core::GlobalState &gs, const core::SymbolData classData,
+                                 const ast::ClassDef *const classDef, const core::SymbolData errMsgClassData,
+                                 const std::string verb) {
+    for (const auto &mixin : classData->mixins()) {
+        if (!mixin.data(gs)->isClassFinal()) {
+            continue;
+        }
+        if (auto e = gs.beginError(getAncestorLoc(gs, classDef, mixin), core::errors::Resolver::FinalAncestor)) {
+            e.setHeader("`{}` was declared as final and cannot be {} by `{}`", mixin.data(gs)->show(gs), verb,
+                        errMsgClassData->show(gs));
+            e.addErrorLine(mixin.data(gs)->loc(), "`{}` defined here", mixin.data(gs)->show(gs));
+        }
+    }
+}
+
+void validateFinalMethodHelper(const core::GlobalState &gs, const core::SymbolData classData,
+                               const core::SymbolData errMsgClassData) {
+    if (!classData->isClassFinal()) {
+        return;
+    }
+    for (const auto mem : classData->members()) {
+        const auto memData = mem.second.data(gs);
+        if (!memData->isMethod() || memData->name == core::Names::staticInit() || memData->isFinalMethod()) {
+            continue;
+        }
+        if (auto e = gs.beginError(memData->loc(), core::errors::Resolver::FinalModuleNonFinalMethod)) {
+            e.setHeader("`{}` was declared as final but its method `{}` was not declared as final",
+                        errMsgClassData->show(gs), memData->name.show(gs));
+        }
+    }
+}
+
+void validateFinal(const core::GlobalState &gs, const core::SymbolData classData, const ast::ClassDef *const classDef) {
```

Comment:
SymbolData is an implementation detail. We shouldn't be passing these around (dmitry and I were both surprised to see that this even compiled—we should probably delete the copy constructor of SymbolData to prevent other people from doing this).

Instead, just take the SymbolRef directly.